lib/perl/Genome/Sys/LockMigrationProxy.t

```diff
@@ -0,0 +1,170 @@
+    plan tests => 4;
+
+    my %backends = (
+        test_old => [ Genome::Sys::Lock::MockBackend->new() ],
+        test_new => [ Genome::Sys::Lock::MockBackend->new() ],
+    );
+    my $backend_guard = backend_guard();
+    my $scopes_guard = Sub::Override->new('Genome::Sys::Lock::scopes', sub { keys %backends });
+    Genome::Sys::Lock->set_backends(%backends);
+
+    my $resource = random_string();
+    my %lock_params = (
+        block_sleep => 0,
+        max_try => 0,
+    );
+
+    my $new_lock = Genome::Sys::Lock->lock_resource(
+        resource_lock => $resource,
+        scope => 'test_new',
+        %lock_params,
+    );
+    ok($new_lock, q(preemptively locked on 'test_new'));
+
+    my $success = Genome::Sys::LockMigrationProxy->new(
+        old => {
+            resource => $resource,
+            scope => 'test_old',
+        },
+        new => {
+            resource => $resource,
+            scope => 'test_new',
+        },
+    )->lock(%lock_params);
+    ok(!$success, q(failed to lock when 'test_new' was already locked));
+
+    my $old_lock = Genome::Sys::Lock->lock_resource(
+        resource_lock => $resource,
+        scope => 'test_old',
+        %lock_params,
+    );
+    ok($old_lock, q(able to lock on 'test_old' after LockMigrationProxy failed to lock));
+
+    $new_lock = Genome::Sys::Lock->lock_resource(
+        resource_lock => $resource,
+        scope => 'test_new',
+        %lock_params,
+    );
+    is($new_lock, undef, q(still locked on 'test_new'));
+};
+
+subtest 'both ok' => sub {
+    plan tests => 3;
+
+    my %backends = (
+        test_old => [ Genome::Sys::Lock::MockBackend->new() ],
+        test_new => [ Genome::Sys::Lock::MockBackend->new() ],
+    );
+    my $backend_guard = backend_guard();
+    my $scopes_guard = Sub::Override->new('Genome::Sys::Lock::scopes', sub { keys %backends });
+    Genome::Sys::Lock->set_backends(%backends);
+
+    my $resource = random_string();
+    my %lock_params = (
+        block_sleep => 0,
+        max_try => 0,
+    );
+
+    my $success = Genome::Sys::LockMigrationProxy->new(
+        old => {
+            resource => $resource,
+            scope => 'test_old',
+        },
+        new => {
+            resource => $resource,
+            scope => 'test_new',
+        },
+    )->lock(%lock_params);
+    ok($success, q(successfully locked LockMigrationProxy));
+
+    my $old_lock = Genome::Sys::Lock->lock_resource(
+        resource_lock => $resource,
+        scope => 'test_old',
+        %lock_params,
+    );
+    is($old_lock, undef, q(failed to lock on 'test_old' after LockMigrationProxy locked));
+
+    my $new_lock = Genome::Sys::Lock->lock_resource(
+        resource_lock => $resource,
+        scope => 'test_new',
+        %lock_params,
+    );
+    is($new_lock, undef, q(failed to lock on 'test_new' after LockMigrationProxy locked));
+};
+
+sub random_string {
+    my @chars = map { (shuffle 'a'..'z', 'A'..'Z', 0..9)[0] } 1..10;
```

Comment:
It's not random, but if we're trying for uniqueness would `Data::UUID->create_str` serve the purpose?

Alternatively, some version of:
```perl
my @valid_chars = ('a'..'z', 'A'..'Z',0..9);
my @chars = map { $valid_chars[rand @valid_chars] } 1..10;
return join('', @chars);
```
should have the same effect without needing to shuffle the entire list each time.